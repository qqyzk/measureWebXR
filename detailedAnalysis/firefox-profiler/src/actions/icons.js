/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type { Action, ThunkAction } from 'firefox-profiler/types';

export function iconHasLoaded(icon: string): Action {
  return {
    type: 'ICON_HAS_LOADED',
    icon,
  };
}

export function iconIsInError(icon: string): Action {
  return {
    type: 'ICON_IN_ERROR',
    icon,
  };
}

const icons: Set<string> = new Set();

type IconRequestResult = 'loaded' | 'error' | 'cached';

function _getIcon(icon: string): Promise<IconRequestResult> {
  if (icons.has(icon)) {
    return Promise.resolve('cached');
  }

  icons.add(icon);

  const result = new Promise((resolve) => {
    const image = new Image();
    image.src = icon;
    image.referrerPolicy = 'no-referrer';
    image.onload = () => {
      resolve('loaded');
    };
    image.onerror = () => {
      resolve('error');
    };
  });

  return result;
}

export function iconStartLoading(icon: string): ThunkAction<Promise<void>> {
  return (dispatch) => {
    return _getIcon(icon).then((result) => {
      switch (result) {
        case 'loaded':
          dispatch(iconHasLoaded(icon));
          break;
        case 'error':
          dispatch(iconIsInError(icon));
          break;
        case 'cached':
          // nothing to do
          break;
        default:
          throw new Error(`Unknown icon load result ${result}`);
      }
    });
  };
}
