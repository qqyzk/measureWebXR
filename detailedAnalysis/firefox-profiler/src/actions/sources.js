/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
import type { Action, SourceLoadingError } from 'firefox-profiler/types';

export function beginLoadingSourceFromUrl(file: string, url: string): Action {
  return { type: 'SOURCE_LOADING_BEGIN_URL', file, url };
}

export function beginLoadingSourceFromBrowserConnection(file: string): Action {
  return { type: 'SOURCE_LOADING_BEGIN_BROWSER_CONNECTION', file };
}

export function finishLoadingSource(file: string, source: string): Action {
  return { type: 'SOURCE_LOADING_SUCCESS', file, source };
}

export function failLoadingSource(
  file: string,
  errors: SourceLoadingError[]
): Action {
  return { type: 'SOURCE_LOADING_ERROR', file, errors };
}
