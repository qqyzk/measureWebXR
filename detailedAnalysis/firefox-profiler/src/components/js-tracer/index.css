/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

.jsTracer {
  display: flex;
  flex: 1;
  flex-flow: column nowrap;
}

.jsTracerLoader {
  display: flex;
  flex: 1;
  flex-flow: column;
  align-items: center;
  justify-content: center;
  border-top: solid 1px var(--grey-20);
  background-color: var(--grey-10);
  font-size: 12px;
  vertical-align: middle;
}

.jsTracerCanvasFadeIn {
  background-color: #fff;
  opacity: 0;
}

.jsTracerCanvasFadeIn.jsTracerCanvasDrawn {
  opacity: 1;
  transition: opacity 200ms;
}
