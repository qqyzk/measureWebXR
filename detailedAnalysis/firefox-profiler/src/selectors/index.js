/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow
export * from './app';
export * from './l10n';
export * from './per-thread';
export * from './profile';
export * from './url-state';
export * from './icons';
export * from './publish';
export * from './zipped-profiles';
export * from './cpu';

import * as app from './app';
import { selectedThreadSelectors as selectedThread } from './per-thread';
import * as profile from './profile';
import * as urlState from './url-state';
import * as icons from './icons';
import * as publish from './publish';
import * as zippedProfiles from './zipped-profiles';
import * as l10n from './l10n';
import * as cpu from './cpu';

const _selectorsForConsole = {
  app,
  profile,
  urlState,
  icons,
  publish,
  zippedProfiles,
  selectedThread,
  l10n,
  cpu,
};

// Exports require explicit typing. Deduce the type with typeof.
export const selectorsForConsole: typeof _selectorsForConsole =
  _selectorsForConsole;
