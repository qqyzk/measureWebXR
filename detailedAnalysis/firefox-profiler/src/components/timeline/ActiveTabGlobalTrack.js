/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import React, { PureComponent } from 'react';
import classNames from 'classnames';
import { selectActiveTabTrack } from 'firefox-profiler/actions/profile-view';
import { getSelectedThreadIndexes } from 'firefox-profiler/selectors/url-state';
import explicitConnect from 'firefox-profiler/utils/connect';
import {
  getActiveTabGlobalTracks,
  getActiveTabResourceTracks,
} from 'firefox-profiler/selectors/profile';
import './Track.css';
import { TimelineTrackThread } from './TrackThread';
import { TimelineTrackScreenshots } from './TrackScreenshots';
import { TimelineActiveTabResourcesPanel } from './ActiveTabResourcesPanel';
import { assertExhaustiveCheck } from 'firefox-profiler/utils/flow';
import { hasThreadKeys } from 'firefox-profiler/profile-logic/profile-data';

import type {
  GlobalTrackReference,
  TrackIndex,
  ActiveTabGlobalTrack,
  InitialSelectedTrackReference,
  ActiveTabResourceTrack,
} from 'firefox-profiler/types';

import type { ConnectedProps } from 'firefox-profiler/utils/connect';

type OwnProps = {|
  +trackReference: GlobalTrackReference,
  +trackIndex: TrackIndex,
  +setInitialSelected: (
    el: InitialSelectedTrackReference,
    forceScroll?: boolean
  ) => void,
|};

type StateProps = {|
  +globalTrack: ActiveTabGlobalTrack,
  +isSelected: boolean,
  +resourceTracks: ActiveTabResourceTrack[],
|};

type DispatchProps = {|
  +selectActiveTabTrack: typeof selectActiveTabTrack,
|};

type Props = ConnectedProps<OwnProps, StateProps, DispatchProps>;

/**
 * Global track of active tab timeline view. Differently from the full view,
 * it shows either screenshot and a single main track for active tab and
 * resources tracks inside it.
 */
class ActiveTabGlobalTrackComponent extends PureComponent<Props> {
  _container: HTMLElement | null = null;
  _isInitialSelectedPane: boolean | null = null;
  _selectCurrentTrack = () => {
    const { selectActiveTabTrack, trackReference } = this.props;
    selectActiveTabTrack(trackReference);
  };

  renderTrack() {
    const { globalTrack } = this.props;
    switch (globalTrack.type) {
      case 'tab': {
        const { threadsKey } = globalTrack;
        return (
          <TimelineTrackThread
            threadsKey={threadsKey}
            showMemoryMarkers={false}
            trackType="expanded"
            trackName="Active Tab"
          />
        );
      }
      case 'screenshots': {
        const { threadIndex, id } = globalTrack;
        return (
          <TimelineTrackScreenshots threadIndex={threadIndex} windowId={id} />
        );
      }
      default:
        console.error(
          'Unhandled active tab globalTrack type',
          (globalTrack: empty)
        );
        return null;
    }
  }

  renderResourcesPanel() {
    const { resourceTracks, setInitialSelected } = this.props;
    if (resourceTracks.length === 0) {
      return null;
    }

    return (
      <TimelineActiveTabResourcesPanel
        resourceTracks={resourceTracks}
        setInitialSelected={setInitialSelected}
      />
    );
  }

  _takeContainerRef = (el: HTMLElement | null) => {
    const { isSelected } = this.props;
    this._container = el;

    if (isSelected) {
      this._isInitialSelectedPane = true;
    }
  };

  componentDidMount() {
    if (this._isInitialSelectedPane && this._container !== null) {
      // Handle the scrolling of the initial selected track into view.
      this.props.setInitialSelected(this._container);
    }
  }

  render() {
    const { isSelected } = this.props;

    return (
      <li ref={this._takeContainerRef} className="timelineTrack">
        <div
          className={classNames(
            'timelineTrackRow timelineTrackGlobalRow activeTab',
            {
              selected: isSelected,
            }
          )}
          onClick={this._selectCurrentTrack}
        >
          <div className="timelineTrackTrack">{this.renderTrack()}</div>
        </div>
        {this.renderResourcesPanel()}
      </li>
    );
  }
}

// Provide an empty list, so that strict equality checks work for component updates.
const EMPTY_RESOURCE_TRACKS = [];

export const TimelineActiveTabGlobalTrack = explicitConnect<
  OwnProps,
  StateProps,
  DispatchProps
>({
  mapStateToProps: (state, { trackIndex }) => {
    const globalTracks = getActiveTabGlobalTracks(state);
    const globalTrack = globalTracks[trackIndex];

    // These get assigned based on the track type.
    let isSelected = false;
    let resourceTracks = EMPTY_RESOURCE_TRACKS;

    // Run different selectors based on the track type.
    switch (globalTrack.type) {
      case 'tab': {
        isSelected = hasThreadKeys(
          getSelectedThreadIndexes(state),
          globalTrack.threadsKey
        );
        resourceTracks = getActiveTabResourceTracks(state);
        break;
      }
      case 'screenshots':
        break;
      default:
        throw assertExhaustiveCheck(
          globalTrack,
          'Unhandled active tab GlobalTrack type.'
        );
    }

    return {
      globalTrack,
      isSelected,
      resourceTracks,
    };
  },
  mapDispatchToProps: {
    selectActiveTabTrack,
  },
  component: ActiveTabGlobalTrackComponent,
});
