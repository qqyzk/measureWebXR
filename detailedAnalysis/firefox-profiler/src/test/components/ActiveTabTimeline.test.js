/* This Source Code Form is subject to the terms of the Mozilla Public
 * License, v. 2.0. If a copy of the MPL was not distributed with this
 * file, You can obtain one at http://mozilla.org/MPL/2.0/. */

// @flow

import * as React from 'react';
import ReactDOM from 'react-dom';
import { Provider } from 'react-redux';

import { render, screen } from 'firefox-profiler/test/fixtures/testing-library';
import { Timeline } from '../../components/timeline';
import { TimelineActiveTabGlobalTrack } from '../../components/timeline/ActiveTabGlobalTrack';
import { TimelineActiveTabResourcesPanel } from '../../components/timeline/ActiveTabResourcesPanel';
import { TimelineActiveTabResourceTrack } from '../../components/timeline/ActiveTabResourceTrack';
import { changeTimelineTrackOrganization } from '../../actions/receive-profile';
import {
  getActiveTabGlobalTracks,
  getActiveTabResourceTracks,
} from '../../selectors/profile';
import { getFirstSelectedThreadIndex } from '../../selectors/url-state';
import { changeSelectedThreads } from '../../actions/profile-view';
import { ensureExists, getFirstItemFromSet } from '../../utils/flow';

import { storeWithProfile } from '../fixtures/stores';
import { getProfileWithNiceTracks } from '../fixtures/profiles/tracks';
import { fireFullClick } from '../fixtures/utils';
import { addActiveTabInformationToProfile } from '../fixtures/profiles/processed-profile';
import { autoMockCanvasContext } from '../fixtures/mocks/canvas-context';
import {
  autoMockElementSize,
  getElementWithFixedSize,
} from '../fixtures/mocks/element-size';
import { mockRaf } from '../fixtures/mocks/request-animation-frame';
import { autoMockIntersectionObserver } from '../fixtures/mocks/intersection-observer';

describe('ActiveTabTimeline', function () {
  autoMockCanvasContext();
  autoMockElementSize({ width: 200, height: 300 });
  autoMockIntersectionObserver();
  beforeEach(() => {
    jest
      .spyOn(ReactDOM, 'findDOMNode')
      .mockImplementation(() =>
        getElementWithFixedSize({ width: 200, height: 300 })
      );
  });

  it('should be rendered properly from the Timeline component', () => {
    const { profile, parentInnerWindowIDsWithChildren, firstTabTabID } =
      addActiveTabInformationToProfile(getProfileWithNiceTracks());
    profile.threads[0].frameTable.innerWindowID[0] =
      parentInnerWindowIDsWithChildren;
    const store = storeWithProfile(profile);
    store.dispatch(
      changeTimelineTrackOrganization({
        type: 'active-tab',
        tabID: firstTabTabID,
      })
    );

    // WithSize uses requestAnimationFrame
    const flushRafCalls = mockRaf();
    const { container } = render(
      <Provider store={store}>
        <Timeline />
      </Provider>
    );
    flushRafCalls();
    expect(container.firstChild).toMatchSnapshot();
  });

  describe('ActiveTabGlobalTrack', function () {
    function setup() {
      const { profile, ...pageInfo } = addActiveTabInformationToProfile(
        getProfileWithNiceTracks()
      );
      profile.threads[0].frameTable.innerWindowID[0] =
        pageInfo.parentInnerWindowIDsWithChildren;
      const store = storeWithProfile(profile);
      store.dispatch(
        changeTimelineTrackOrganization({
          type: 'active-tab',
          tabID: pageInfo.firstTabTabID,
        })
      );
      const trackIndex = 0;
      const { getState, dispatch } = store;
      const trackReference = { type: 'global', trackIndex };
      const tracks = getActiveTabGlobalTracks(getState());
      const track = tracks[trackIndex];
      const setInitialSelected = () => {};
      if (track.type !== 'tab') {
        throw new Error('Expected a tab track.');
      }
      const threadIndex = ensureExists(
        getFirstItemFromSet(track.threadIndexes),
        'Expected a thread index for given active tab global track'
      );

      // The assertions are simpler if the GeckoMain tab thread is not already selected.
      dispatch(changeSelectedThreads(new Set([threadIndex + 1])));

      // WithSize uses requestAnimationFrame
      const flushRafCalls = mockRaf();

      const renderResult = render(
        <Provider store={store}>
          <TimelineActiveTabGlobalTrack
            trackIndex={trackIndex}
            trackReference={trackReference}
            setInitialSelected={setInitialSelected}
          />
        </Provider>
      );
      flushRafCalls();

      const { container } = renderResult;

      const getGlobalTrackRow = () =>
        ensureExists(
          container.querySelector('.timelineTrackGlobalRow'),
          `Couldn't find the track global row with selector .timelineTrackGlobalRow`
        );

      return {
        ...renderResult,
        ...pageInfo,
        dispatch,
        getState,
        profile,
        store,
        trackReference,
        trackIndex,
        threadIndex,
        getGlobalTrackRow,
      };
    }

    it('matches the snapshot of a global tab track', () => {
      const { container } = setup();
      expect(container.firstChild).toMatchSnapshot();
    });

    it('has useful parts of the component', function () {
      const { getGlobalTrackRow } = setup();
      expect(getGlobalTrackRow()).toBeTruthy();
    });

    it('starts out not being selected', function () {
      const { getState, getGlobalTrackRow, threadIndex } = setup();
      expect(getFirstSelectedThreadIndex(getState())).not.toBe(threadIndex);
      expect(getGlobalTrackRow()).not.toHaveClass('selected');
    });

    it('can select a thread by clicking the row', () => {
      const { getState, getGlobalTrackRow, threadIndex } = setup();
      expect(getFirstSelectedThreadIndex(getState())).not.toBe(threadIndex);
      fireFullClick(getGlobalTrackRow());
      expect(getFirstSelectedThreadIndex(getState())).toBe(threadIndex);
    });

    it('does not display the resources panel if there are no resource tracks', () => {
      const { getState } = setup();
      expect(getActiveTabResourceTracks(getState()).length).toBe(0);
      expect(screen.queryByText(/Resources/)).not.toBeInTheDocument();
    });
  });

  describe('ActiveTabResourcesPanel', function () {
    function setup() {
      const { profile, ...pageInfo } = addActiveTabInformationToProfile(
        getProfileWithNiceTracks()
      );
      const mainThreadIndex = 0;
      const resourceThreadIndex = 1;
      // Setting the first thread as parent track and the second as the iframe track.
      profile.threads[mainThreadIndex].frameTable.innerWindowID[0] =
        pageInfo.parentInnerWindowIDsWithChildren;
      profile.threads[resourceThreadIndex].frameTable.innerWindowID[0] =
        pageInfo.iframeInnerWindowIDsWithChild;
      const store = storeWithProfile(profile);
      store.dispatch(
        changeTimelineTrackOrganization({
          type: 'active-tab',
          tabID: pageInfo.firstTabTabID,
        })
      );
      const { getState, dispatch } = store;
      const resourceTracks = getActiveTabResourceTracks(getState());

      // WithSize uses requestAnimationFrame
      const flushRafCalls = mockRaf();

      const renderResult = render(
        <Provider store={store}>
          <TimelineActiveTabResourcesPanel
            resourceTracks={resourceTracks}
            setInitialSelected={() => {}}
          />
        </Provider>
      );

      flushRafCalls();

      const getResourcesPanelHeader = () => screen.getByText(/Resources/);
      const getResourceFrameTrack = () => screen.queryByText(/IFrame:/);

      return {
        ...renderResult,
        ...pageInfo,
        dispatch,
        getState,
        profile,
        store,
        getResourcesPanelHeader,
        getResourceFrameTrack,
        mainThreadIndex,
        resourceThreadIndex,
        flushRafCalls,
      };
    }

    it('matches the snapshot of a resources panel when closed', () => {
      const { container } = setup();
      expect(container.firstChild).toMatchSnapshot();
    });

    it('matches the snapshot of a resources panel when opened', () => {
      const { container, getResourcesPanelHeader, flushRafCalls } = setup();
      fireFullClick(getResourcesPanelHeader());
      flushRafCalls();
      expect(container.firstChild).toMatchSnapshot();
    });

    it('is closed by default', () => {
      const { getResourceFrameTrack } = setup();
      expect(getResourceFrameTrack()).toBeFalsy();
    });

    it('clicking on the header opens the resources panel', () => {
      const { getResourcesPanelHeader, getResourceFrameTrack, flushRafCalls } =
        setup();
      const resourcesPanelHeader = getResourcesPanelHeader();
      expect(getResourceFrameTrack()).toBeFalsy();

      fireFullClick(resourcesPanelHeader);
      flushRafCalls();
      expect(getResourceFrameTrack()).toBeTruthy();
    });

    it('selects the main track when panel is being closed', () => {
      const {
        getResourcesPanelHeader,
        getResourceFrameTrack,
        getState,
        mainThreadIndex,
        resourceThreadIndex,
      } = setup();
      // At first, make sure the main thread is selected.
      expect(getFirstSelectedThreadIndex(getState())).toBe(mainThreadIndex);

      // 1. Open the panel.
      fireFullClick(getResourcesPanelHeader());
      // 2. Select the reource track.
      fireFullClick(ensureExists(getResourceFrameTrack()));
      // Selected thread should be the resource now.
      expect(getFirstSelectedThreadIndex(getState())).toBe(resourceThreadIndex);

      // 3. Close the panel.
      fireFullClick(getResourcesPanelHeader());
      // Now the main thread should be selected again.
      expect(getFirstSelectedThreadIndex(getState())).toBe(mainThreadIndex);
    });
  });

  describe('ActiveTabResourceTrack', function () {
    function setup() {
      const { profile, ...pageInfo } = addActiveTabInformationToProfile(
        getProfileWithNiceTracks()
      );
      // Setting the threads with the following relationship:
      // Page #1 (thread #0)
      // |- Page #2 (thread #1)
      //    |- Page #3 (thread #2)
      const threadIndex = 2;
      profile.threads[0].frameTable.innerWindowID[0] =
        pageInfo.parentInnerWindowIDsWithChildren;
      profile.threads[1].frameTable.innerWindowID[0] =
        pageInfo.iframeInnerWindowIDsWithChild;
      profile.threads[threadIndex].frameTable.innerWindowID[0] =
        pageInfo.firstTabInnerWindowIDs[2];
      profile.threads[threadIndex].name = 'GeckoMain';
      profile.threads[threadIndex].isMainThread = true;
      const store = storeWithProfile(profile);
      store.dispatch(
        changeTimelineTrackOrganization({
          type: 'active-tab',
          tabID: pageInfo.firstTabTabID,
        })
      );
      const { getState, dispatch } = store;
      const resourceTracks = getActiveTabResourceTracks(getState());
      const trackIndex = 1;

      // WithSize uses requestAnimationFrame
      const flushRafCalls = mockRaf();
      const renderResult = render(
        <Provider store={store}>
          <TimelineActiveTabResourceTrack
            resourceTrack={resourceTracks[1]}
            trackIndex={trackIndex}
            setInitialSelected={() => {}}
          />
        </Provider>
      );
      flushRafCalls();

      const { container } = renderResult;
      const resourcePage = ensureExists(profile.pages)[2];
      const getResourceFrameTrackLabel = () =>
        screen.getByText(resourcePage.url);
      const getResourceTrackRow = () =>
        ensureExists(
          container.querySelector('.timelineTrackResourceRow'),
          `Couldn't find the track resource row with selector .timelineTrackResourceRow`
        );
      const isResourceTrackOpen = () =>
        getResourceTrackRow().classList.contains('opened');

      return {
        ...renderResult,
        ...pageInfo,
        dispatch,
        getState,
        profile,
        store,
        threadIndex,
        getResourceFrameTrackLabel,
        getResourceTrackRow,
        resourcePage,
        isResourceTrackOpen,
      };
    }

    describe('with a thread/sub-frame track', function () {
      it('matches the snapshot of a resource track', () => {
        const { container } = setup();
        expect(container.firstChild).toMatchSnapshot();
      });

      it('has the correct track name', function () {
        const { resourcePage } = setup();
        expect(screen.getByText(resourcePage.url)).toBeInTheDocument();
      });

      it('starts out not being selected', function () {
        const { getState, threadIndex } = setup();
        expect(getFirstSelectedThreadIndex(getState())).not.toBe(threadIndex);
      });

      it('can select a thread by clicking the label', () => {
        const { getState, getResourceFrameTrackLabel, threadIndex } = setup();
        expect(getFirstSelectedThreadIndex(getState())).not.toBe(threadIndex);
        fireFullClick(getResourceFrameTrackLabel());
        expect(getFirstSelectedThreadIndex(getState())).toBe(threadIndex);
      });

      it('can select a thread by clicking the row', () => {
        const { getState, getResourceTrackRow, threadIndex } = setup();
        expect(getFirstSelectedThreadIndex(getState())).not.toBe(threadIndex);
        fireFullClick(getResourceTrackRow());
        expect(getFirstSelectedThreadIndex(getState())).toBe(threadIndex);
      });

      it('can toggle a selected track by clicking the label', () => {
        const {
          getState,
          getResourceFrameTrackLabel,
          threadIndex,
          isResourceTrackOpen,
        } = setup();
        expect(getFirstSelectedThreadIndex(getState())).not.toBe(threadIndex);
        expect(isResourceTrackOpen()).toBe(false);
        fireFullClick(getResourceFrameTrackLabel());
        expect(getFirstSelectedThreadIndex(getState())).toBe(threadIndex);
        expect(isResourceTrackOpen()).toBe(true);
      });

      it('does not toggle a selected track by clicking other part of the track except label', () => {
        const {
          getState,
          getResourceTrackRow,
          threadIndex,
          isResourceTrackOpen,
        } = setup();
        expect(getFirstSelectedThreadIndex(getState())).not.toBe(threadIndex);
        expect(isResourceTrackOpen()).toBe(false);
        fireFullClick(getResourceTrackRow());
        expect(getFirstSelectedThreadIndex(getState())).toBe(threadIndex);
        expect(isResourceTrackOpen()).toBe(false);
      });
    });
  });
});
