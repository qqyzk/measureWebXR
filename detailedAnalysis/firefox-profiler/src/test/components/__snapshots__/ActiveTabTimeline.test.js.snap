// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ActiveTabTimeline ActiveTabGlobalTrack matches the snapshot of a global tab track 1`] = `
<li
  class="timelineTrack"
>
  <div
    class="timelineTrackRow timelineTrackGlobalRow activeTab"
  >
    <div
      class="timelineTrackTrack"
    >
      <div
        class="timelineTrackThread expanded"
      >
        <div
          class="timelineMarkers"
          data-testid="TimelineMarkersJank"
        >
          <div
            class="react-contextmenu-wrapper"
          >
            <div>
              <canvas
                class="timelineMarkersCanvas"
                height="300"
                width="200"
              />
            </div>
          </div>
        </div>
        <div
          class="timelineMarkers timelineMarkersGeckoMain"
          data-testid="TimelineMarkersOverview"
        >
          <div
            class="react-contextmenu-wrapper"
          >
            <div>
              <canvas
                class="timelineMarkersCanvas"
                height="300"
                width="200"
              />
            </div>
          </div>
        </div>
        <div
          class="threadActivityGraph"
        >
          <div>
            <canvas
              class="threadActivityGraphCanvas threadActivityGraphCanvas"
              height="300"
              width="200"
            >
              <h2>
                Activity Graph for 
                Active Tab
              </h2>
              <p>
                This graph shows a visual chart of thread activity.
              </p>
            </canvas>
          </div>
        </div>
        <div
          class="threadSampleGraph"
        >
          <div>
            <canvas
              class="threadSampleGraphCanvas threadSampleGraphCanvas"
              height="300"
              width="200"
            >
              <h2>
                Stack Graph for 
                Active Tab
              </h2>
              <p>
                This graph charts the stack height of each sample.
              </p>
            </canvas>
          </div>
        </div>
        <div
          class="timelineEmptyThreadIndicator"
        />
      </div>
    </div>
  </div>
</li>
`;

exports[`ActiveTabTimeline ActiveTabResourceTrack with a thread/sub-frame track matches the snapshot of a resource track 1`] = `
<li
  class="timelineTrack timelineTrackResource"
>
  <div
    class="timelineTrackRow timelineTrackResourceRow"
  >
    <div
      class="timelineTrackResourceLabel"
    >
      <span>
        IFrame:
      </span>
       
      Page #3
    </div>
    <div
      class="timelineTrackTrack"
    >
      <div
        class="timelineTrackThread condensed"
      >
        <div
          class="timelineMarkers"
          data-testid="TimelineMarkersJank"
        >
          <div
            class="react-contextmenu-wrapper"
          >
            <div>
              <canvas
                class="timelineMarkersCanvas"
                height="300"
                width="200"
              />
            </div>
          </div>
        </div>
        <div
          class="threadActivityGraph"
        >
          <div>
            <canvas
              class="threadActivityGraphCanvas threadActivityGraphCanvas"
              height="300"
              width="200"
            >
              <h2>
                Activity Graph for 
                Page #3
              </h2>
              <p>
                This graph shows a visual chart of thread activity.
              </p>
            </canvas>
          </div>
        </div>
        <div
          class="timelineEmptyThreadIndicator"
        />
      </div>
    </div>
  </div>
</li>
`;

exports[`ActiveTabTimeline ActiveTabResourcesPanel matches the snapshot of a resources panel when closed 1`] = `
<div
  class="timelineResources"
  style="--resources-header-height: 20px;"
>
  <div
    class="timelineResourcesHeader"
  >
    Resources (
    1
    )
    <div
      class="timelineTrackThread condensed"
    >
      <div
        class="timelineMarkers"
        data-testid="TimelineMarkersJank"
      >
        <div
          class="react-contextmenu-wrapper"
        >
          <div>
            <canvas
              class="timelineMarkersCanvas"
              height="300"
              width="200"
            />
          </div>
        </div>
      </div>
      <div
        class="threadActivityGraph"
      >
        <div>
          <canvas
            class="threadActivityGraphCanvas threadActivityGraphCanvas"
            height="300"
            width="200"
          >
            <h2>
              Activity Graph for 
              Merged resource tracks
            </h2>
            <p>
              This graph shows a visual chart of thread activity.
            </p>
          </canvas>
        </div>
      </div>
      <div
        class="timelineEmptyThreadIndicator"
      />
    </div>
  </div>
</div>
`;

exports[`ActiveTabTimeline ActiveTabResourcesPanel matches the snapshot of a resources panel when opened 1`] = `
<div
  class="timelineResources"
  style="--resources-header-height: 20px;"
>
  <div
    class="timelineResourcesHeader opened"
  >
    Resources (
    1
    )
    <div
      class="timelineTrackThread condensed"
    >
      <div
        class="timelineMarkers"
        data-testid="TimelineMarkersJank"
      >
        <div
          class="react-contextmenu-wrapper"
        >
          <div>
            <canvas
              class="timelineMarkersCanvas"
              height="300"
              width="200"
            />
          </div>
        </div>
      </div>
      <div
        class="threadActivityGraph"
      >
        <div>
          <canvas
            class="threadActivityGraphCanvas threadActivityGraphCanvas"
            height="300"
            width="200"
          >
            <h2>
              Activity Graph for 
              Merged resource tracks
            </h2>
            <p>
              This graph shows a visual chart of thread activity.
            </p>
          </canvas>
        </div>
      </div>
      <div
        class="timelineEmptyThreadIndicator"
      />
    </div>
  </div>
  <ol
    class="timelineResourceTracks"
  >
    <li
      class="timelineTrack timelineTrackResource"
    >
      <div
        class="timelineTrackRow timelineTrackResourceRow"
      >
        <div
          class="timelineTrackResourceLabel"
        >
          <span>
            IFrame:
          </span>
           
          Page #2
        </div>
        <div
          class="timelineTrackTrack"
        >
          <div
            class="timelineTrackThread condensed"
          >
            <div
              class="timelineMarkers"
              data-testid="TimelineMarkersJank"
            >
              <div
                class="react-contextmenu-wrapper"
              >
                <div>
                  <canvas
                    class="timelineMarkersCanvas"
                    height="300"
                    width="200"
                  />
                </div>
              </div>
            </div>
            <div
              class="threadActivityGraph"
            >
              <div>
                <canvas
                  class="threadActivityGraphCanvas threadActivityGraphCanvas"
                  height="300"
                  width="200"
                >
                  <h2>
                    Activity Graph for 
                    Page #2
                  </h2>
                  <p>
                    This graph shows a visual chart of thread activity.
                  </p>
                </canvas>
              </div>
            </div>
            <div
              class="timelineEmptyThreadIndicator"
            />
          </div>
        </div>
      </div>
    </li>
  </ol>
</div>
`;

exports[`ActiveTabTimeline should be rendered properly from the Timeline component 1`] = `
<div
  class="timelineSelection activeTab"
>
  <div
    class="timelineRuler"
    style="--timeline-ruler-height: 20px;"
  >
    <ol
      class="timelineRulerContainer"
    >
      <li
        class="timelineRulerNotch"
        style="left: 0px;"
      >
        <span
          class="timelineRulerNotchText"
        >
          0.0000s
        </span>
      </li>
      <li
        class="timelineRulerNotch"
        style="left: 100px;"
      >
        <span
          class="timelineRulerNotchText"
        >
          0.0005s
        </span>
      </li>
      <li
        class="timelineRulerNotch"
        style="left: 200px;"
      >
        <span
          class="timelineRulerNotchText"
        >
          0.0010s
        </span>
      </li>
    </ol>
  </div>
  <div
    class="overflowEdgeIndicator tracksContainer timelineOverflowEdgeIndicator"
  >
    <div
      class="overflowEdgeIndicatorEdge topEdge"
    />
    <div
      class="overflowEdgeIndicatorEdge rightEdge"
    />
    <div
      class="overflowEdgeIndicatorEdge bottomEdge"
    />
    <div
      class="overflowEdgeIndicatorEdge leftEdge"
    />
    <div
      class="overflowEdgeIndicatorScrollbox tracksContainer timelineOverflowEdgeIndicatorScrollbox"
    >
      <div
        class="overflowEdgeIndicatorContentsWrapper"
      >
        <ol
          class="timelineThreadList"
        >
          <li
            class="timelineTrack"
          >
            <div
              class="timelineTrackRow timelineTrackGlobalRow activeTab selected"
            >
              <div
                class="timelineTrackTrack"
              >
                <div
                  class="timelineTrackThread expanded"
                >
                  <div
                    class="timelineMarkers selected"
                    data-testid="TimelineMarkersJank"
                  >
                    <div
                      class="react-contextmenu-wrapper"
                    >
                      <div>
                        <canvas
                          class="timelineMarkersCanvas"
                          height="300"
                          width="200"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="timelineMarkers timelineMarkersGeckoMain selected"
                    data-testid="TimelineMarkersOverview"
                  >
                    <div
                      class="react-contextmenu-wrapper"
                    >
                      <div>
                        <canvas
                          class="timelineMarkersCanvas"
                          height="300"
                          width="200"
                        />
                      </div>
                    </div>
                  </div>
                  <div
                    class="threadActivityGraph"
                  >
                    <div>
                      <canvas
                        class="threadActivityGraphCanvas threadActivityGraphCanvas"
                        height="300"
                        width="200"
                      >
                        <h2>
                          Activity Graph for 
                          Active Tab
                        </h2>
                        <p>
                          This graph shows a visual chart of thread activity.
                        </p>
                      </canvas>
                    </div>
                  </div>
                  <div
                    class="threadSampleGraph"
                  >
                    <div>
                      <canvas
                        class="threadSampleGraphCanvas threadSampleGraphCanvas"
                        height="300"
                        width="200"
                      >
                        <h2>
                          Stack Graph for 
                          Active Tab
                        </h2>
                        <p>
                          This graph charts the stack height of each sample.
                        </p>
                      </canvas>
                    </div>
                  </div>
                  <div
                    class="timelineEmptyThreadIndicator"
                  />
                </div>
              </div>
            </div>
          </li>
        </ol>
      </div>
    </div>
  </div>
  <div
    class="timelineSelectionHoverLine"
    style="visibility: hidden; left: 0px;"
  />
</div>
`;
